### ISSUE TITLE
oci_load_balancer_backend not removed on port change.

### ISSUE THREAD

**Issue by nshttpd** [oracle/terraform-provider-oci#519]
### Terraform Version

<details>
Terraform v0.11.7
+ provider.oci v2.1.5
</details>

### Description:

When changing out the port on a `oci_load_balancer_backend` the provider says that it's a destroy and recreate of the backend on/in the `oci_load_balancer_backendset`. When applied the new backend is there as part of the backendset, but the old one using the original port is also there. 

### Example:

Setup a backend on port 8081. Change the port to 8080. The TF plan will say it's a destroy and rebuild. After apply look in the UI Console and both the 8081 and 8080 Backend will be associated with the backendset.

**Comment by parweza** [comment]
Hi,
Can you please tf config that you are using as I am not able to reproduce the issue using  following config ( created load balancer with single backend [lb_full](https://github.com/oracle/terraform-provider-oci/blob/master/docs/examples/load_balancer/lb_full/lb_full.tf)
Please note that the example mentioned above has two lb backend with port 80.

**Comment by nshttpd** [comment]
In this case I took the `oci_load_balancer_backend` and moved it from 8081 to port 80 and did an apply. The old backend and new backend were both there in the backendset. If you want to see all the TF I can ping you on internal email for access to ALM repo.

```
resource "oci_load_balancer_backendset" "yum_backend" {
  load_balancer_id = "${data.terraform_remote_state.vcn_shared.ssvcs_ext_lb_ocid}"
  name             = "yum-backend"
  policy           = "LEAST_CONNECTIONS"

  health_checker {
    interval_ms         = 15000
    port                = 8081
    protocol            = "HTTP"
    retries             = 3
    return_code         = 200
    response_body_regex = "OK"
    url_path            = "/artifactory/api/system/ping"
  }
}

resource "oci_load_balancer_backend" "yum_backend" {
  depends_on       = ["oci_load_balancer_backendset.yum_backend"]
  load_balancer_id = "${data.terraform_remote_state.vcn_shared.ssvcs_ext_lb_ocid}"
  backendset_name  = "yum-backend"
  ip_address       = "${oci_core_instance.repo_01.private_ip}"
  port             = 8081
  backup           = false
  drain            = false
  offline          = false
  weight           = 1
}

resource "oci_load_balancer_listener" "yum" {
  depends_on               = ["oci_load_balancer_backendset.yum_backend"]
  load_balancer_id         = "${data.terraform_remote_state.vcn_shared.ssvcs_ext_lb_ocid}"
  name                     = "yum-ssl-listener"
  default_backend_set_name = "yum-backend"
  port                     = 443
  protocol                 = "HTTP"

  ssl_configuration {
    certificate_name        = "yum-iad-awesome-com"
    verify_peer_certificate = "false"
  }
}
```

**Comment by parweza** [comment]
I was able to reproduce the issue with following configuration and updating both backend set and backend port
[lb_full.txt](https://github.com/oracle/terraform-provider-oci/files/1996429/lb_full.txt)

**Comment by kohashim** [comment]
I was able to reproduce only when _both_ 1) the port of the health_checker in a backendset and 2) the port of a backend at the same time.  When I changed only 2), I was not able to reproduce.  Investigating further.

**Comment by kohashim** [comment]
@nshttpd, can you try "terraform apply -parallelism=1"?  This trick worked consistently for me.

In the meantime, I will engage the load balancer service team to investigate this issue.
Without this parameter, Terraform called 1) UpdateBackendSet API for health check change and 2) DeleteBackend API for deleting the backend with the original port concurrently, then CreateBackend API for creating a new backend with the new port number.  Somehow this API calling pattern revived the deleted backend without calling CreateBackend API.  Most likely this is caused by the service side because I didn't see any evidence of CreateBackend API call made by Terraform.

**Comment by kohashim** [comment]
We found the root cause in the OCI terraform provider, not on the service side.

UpdateBackendSet API accepts UpdateBackendSetDetails body: https://docs.us-phoenix-1.oraclecloud.com/api/#/en/loadbalancer/20170115/datatypes/UpdateBackendSetDetails, which requires to include all attributes, not only the attributes you want to update, but also the other existing attributes.  In other words, the update API is not a patch API.

When both a backendset and a backend that belongs to the backendset are updated in your TF config, Terraform 1) updates the backendset and 2) deletes the backend concurrently as long as the parallelism > 1.  UpdateBackendSetDetails includes the backend it is trying to delete whereas DeleteBackend API is deleting the backend.  As a result, the service first deletes the backend upon the DeleteBackend API request, then recreates the backend again upon UpdateBackendSet API request.

Thus, you can face this issue when you update not only port but any attributes in a backendset and its backend(s) atomically in your TF config.  I confirmed.

We are currently fixing this concurrency bug.

**Comment by alexng-canuck** [comment]
Hello @nshttpd , 

Just wanted to follow-up on whether @kohashim 's analysis explains the issue you were seeing. This issue would happen if you change the config of both the backend and backend_set resources. From your initial description, were you also making changes to both resources before hitting this?

**Comment by nshttpd** [comment]
@alexng-canuck in this case only the `oci_load_balancer_backend` resource was being changed. No change was initiated on the `backend_set`.

**Comment by rcohenma** [comment]
This issue was fixed in release 2.1.12
https://github.com/oracle/terraform-provider-oci/releases/tag/v2.1.12
